components/swap/SwapTokenChart.tsx

```diff
@@ -165,6 +185,113 @@ const SwapTokenChart = () => {
   const swapHistory = mangoStore((s) => s.mangoAccount.swapHistory.data)
   const loadSwapHistory = mangoStore((s) => s.mangoAccount.swapHistory.loading)
   const [showSwaps, setShowSwaps] = useState(true)
+  const [swapTooltipData, setSwapTooltipData] =
+    useState<SwapHistoryItem | null>(null)
+  const [swapTooltipCoingeckoPrice, setSwapTooltipCoingeckoPrice] = useState<
+    string | number | undefined
+  >(undefined)
+
+  const handleSwapMouseEnter = (
+    swap: SwapHistoryItem | undefined,
+    coingeckoPrice: string | number | undefined
+  ) => {
+    if (swap) {
+      setSwapTooltipData(swap)
+    }
+    if (coingeckoPrice) {
+      setSwapTooltipCoingeckoPrice(coingeckoPrice)
+    }
+  }
+
+  const handleSwapMouseLeave = () => {
+    setSwapTooltipData(null)
+  }
+
+  const renderTooltipContent = (swap: SwapHistoryItem) => {
+    const {
```

Comment:
everything in this component should probably have a `useMemo` or `useCallback` since it is especially prone to frequent rerenders